File: 2.2/haskell/RelaxLoop22.hs
module Main where

import System.CPUTime (getCPUTime)

iters :: Int
iters = 20000000

step :: Double -> Double -> Double -> Double
step old du w =
  let alt = du + w
   in if alt < old then alt else old

loop :: Int -> Double -> Double -> Double -> Double -> Double
loop 0 acc _ _ _ = acc
loop k acc old du w =
  let old' = step old du w
      du' = if du + 0.000001 > 1000.0 then 1.0 else du + 0.000001
      w' = if w + 0.000002 > 1000.0 then 0.5 else w + 0.000002
   in loop (k - 1) (acc + old') old' du' w'

main :: IO ()
main = do
  t0 <- getCPUTime  -- picoseconds
  let acc = loop iters 0.0 1000000.0 1.0 0.5
  t1 <- getCPUTime
  let ms = fromIntegral (t1 - t0) / 1e6  -- picoseconds to milliseconds
  putStrLn $ "time_ms: " ++ show ms
  putStrLn $ "checksum: " ++ show acc
  putStrLn $ "iters: " ++ show iters

File: 2.2/rust/relax_loop_22.rs
use std::time::Instant;

const ITERS: u64 = 100_000_000;

#[inline(always)]
fn step(old: f64, du: f64, w: f64) -> f64 {
    let alt = du + w;
    if alt < old {
        alt
    } else {
        old
    }
}

fn main() {
    let mut acc = 0.0f64;
    let mut old = 1_000_000.0f64;
    let mut du = 1.0f64;
    let mut w = 0.5f64;

    let t0 = Instant::now();
    for _ in 0..ITERS {
        old = step(old, du, w);
        du += 0.000001;
        w += 0.000002;
        acc += old;
        if du > 1000.0 {
            du = 1.0;
        }
        if w > 1000.0 {
            w = 0.5;
        }
    }
    let ms = t0.elapsed().as_secs_f64() * 1000.0;

    println!("time_ms: {:.2}", ms);
    println!("checksum: {:.6}", acc);
    println!("iters: {}", ITERS);
}
